decryptor/postgresql/packet_handler.go

```diff
@@ -427,100 +444,159 @@ const WithoutMessageType = 0
+// | Flush               | F    | Byte1('H') || Int32(len)      |
+// | FunctionCall        | F    | Byte1('F') || Int32(len)      |
+// | GSSENCRequest       | F    | Int32(8)   || Int32(80877104) |
+// | GSSResponse         | F    | Byte1('p') || Int32(len)      |
+// | Parse               | F    | Byte1('P') || Int32(len)      |
+// | PasswordMessage     | F    | Byte1('p') || Int32(len)      |
+// | Query               | F    | Byte1('Q') || Int32(len)      |
+// | SASLInitialResponse | F    | Byte1('p') || Int32(len)      |
+// | SASLResponse        | F    | Byte1('p') || Int32(len)      |
+// | SSLRequest          | F    | Int32(8)   || Int32(80877103) |
+// | StartupMessage      | F    | Int32(len) || Int32(196608)   |
+// | Sync                | F    | Byte1('S') || Int32(len)      |
+// | Terminate           | F    | Byte1('X') || Int32(len)      |
+// ```
+//
+// Startup message can only be received as first message after connection is
+// established. This fact allows to distinguish which format to parse.
+// If connection is established, and first message is not startup one, this
+// function returns ErrUnsupportedPacketType.
+// If startup message is already received, but then comes a message with unknown
+// format, it would be parsed as a general message with unknown tag.
+//
+// https://www.postgresql.org/docs/current/static/protocol-message-formats.html
 func (packet *PacketHandler) ReadClientPacket() error {
+	if packet.started {
+		return packet.readGeneralPacket()
+	}
+	// https://www.postgresql.org/docs/current/protocol-flow.html
+	// First packet should always be StartupMessage/SSLRequest/GSSENCRequest
+	return packet.readStartupPacket()
+}
+
+// Tries to read one of the startup packets:
+// - StartupMessage
+// - SSLRequest
+// - CancelRequest
+// - GSSENCRequest
+//
+// Due to historical reasons, all startup messages have the following format:
+//
+//     [4-byte length] [4-byte tag] [payload...]
+//
+// If the packet cannot be parsed as a startup packets, the ErrUnsupportedPacketType
+// is returned.
+//
+// Source: https://www.postgresql.org/docs/current/protocol-flow.html
+func (packet *PacketHandler) readStartupPacket() error {
 	packet.Reset()
-	// 8 bytes because startup/ssl/cancel messages has at least 8 bytes
+	// 8 bytes because all startup messages has at least 8 bytes
 	packetBuf := make([]byte, 8)
 	packet.messageType[0] = WithoutMessageType
-	// any message has at least 5 bytes: TypeOfMessage(1) + Length(4) or 8 bytes of special messages
+
+	n, err := io.ReadFull(packet.reader, packetBuf)
+	if err := base.CheckReadWrite(n, 8, err); err != nil {
+		packet.logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorDecryptorReadPacket).WithError(err).Debugln("Can't read 8 bytes of Startup/SSLRequest packet")
+		return err
+	}
+
+	switch {
+	case bytes.Equal(StartupRequest, packetBuf[4:8]):
+	case bytes.Equal(SSLRequestHeader, packetBuf[:8]):
+	case bytes.Equal(CancelRequestHeader, packetBuf[:8]):
```

Comment:
actually Cancel request is not as general, but it cannot be first. And we should parse it in general process flow.

Reply:
From the [docs](https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.9):
> During the processing of a query, the frontend might request cancellation of the query. The cancel request is not sent directly on the open connection to the backend for reasons of implementation efficiency: we don't want to have the backend constantly checking for new input from the frontend during query processing. Cancel requests should be relatively infrequent, so we make them slightly cumbersome in order to avoid a penalty in the normal case.
To issue a cancel request, the frontend opens a new connection to the server and sends a CancelRequest message, rather than the StartupMessage message that would ordinarily be sent across a new connection. The server will process this request and then close the connection. For security reasons, no direct reply is made to the cancel request message.

So, I thought it's one of the _startup_ messages. That also explains why `CancelRequest` doesn't have general format and uses the same format as the `StartupMessage`.

Reply:
yes, but it is during processing a query. But query they can process only after Startup/SSL and authentication